packages/fluentui/react-bindings/src/hooks/useCallbackRef.ts

```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
+
+/**
+ * Creates a MutableRef with ref change callback. Is useful as React.useRef() doesn't notify you when its content
+ * changes and mutating the .current property doesn't cause a re-render. An opt-out will be use a callback ref via
+ * React.useState(), but it will cause re-renders always.
+ *
+ * @param initialValue - initial ref value
+ * @param {Function} callback - a callback to run when value changes
+ * @param {Boolean} skipInitialResolve - a flag to skip an initial ref report
+ *
+ * @example
+ * const ref = useCallbackRef(0, (newValue, oldValue) => console.log(oldValue, '->', newValue);
+ * ref.current = 1;
+ * // prints 0 -> 1
+ *
+ * @see https://reactjs.org/docs/hooks-reference.html#useref
+ * @see https://github.com/theKashey/use-callback-ref#usecallbackref---to-replace-reactuseref
+ * @returns {MutableRefObject}
+ */
+export function useCallbackRef<T>(
+  initialValue: T | null,
+  callback: (newValue: T | null, lastValue: T | null) => void,
+  skipInitialResolve?: boolean,
+): React.MutableRefObject<T | null> {
+  const isFirst = React.useRef(true);
+  const [ref] = React.useState(() => ({
```

Comment:
I got very curious about this `useCallbackRef` so I have some question, please illuminate me 😁

Why do we use `useState` there if it's never updated ( There's no usage of the dispatch, wouldn't that be the same as just keeping a variable there?) ?

The facade prop would be a good usage of Proxy ? Something on these lines:

https://codesandbox.io/s/fervent-mcnulty-jqomz?file=/src/useCallbackRefProxy.js

Reply:
use state without the dispatch is a great way of doing a 'singleton' in React since you can use a factory function to initialize value.

> Proxy

It could work... but has zero IE11 support, mobx moved to `Proxy` a while ago and basically killed all IE11 support for that lib forever more it wasn't a pretty time :/ 

> Something on these lines:

You would still need a `useState` or `useRef` to store the object since your example doesn't handle rerenders. I cloned that sandbox and added a rerender button... you'll see that referentially in a rerender your `ref` object gets recreated whenever the component rerenders again

Reply:
https://codesandbox.io/s/trusting-brook-s9s5z?file=/src/useCallbackRefProxy.js

Reply:
Thanks for clarification ❤️

Reply:
_Why not `useMemo()`?_

> useMemo() is a hint — not a guarantee. React may still choose to “forget” some memoized values to reclaim memory. Don’t rely on it for correctness! 

https://twitter.com/dan_abramov/status/1085685515951714306?s=20